### PR TITLE
Partial protocol selection

### PR DESCRIPTION
**Pull request by rolph-recto**
This modifies protocol selection to respect partial protocol assignments with which a program has been annotated. This also modifies the `compile` command to skip protocol selection if all let nodes and declarations are already annotated with protocols. This is useful for code generation of annotated intermediate programs; you don't need to run protocol selection every time you want to generate code.

### CODE REVIEW

**Comment by cacay** [comment]
Why does this have to explicitly check if the program is fully annotated and skip selection? If it is fully annotated, we have a trivial selection which Z3 should immediately solve.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/SelectionConstraintGenerator.kt [hunk 1]
```diff
@@ -62,17 +62,17 @@ class SelectionConstraintGenerator(
 
     // TODO: pc must be weak enough for the hosts involved in the selected protocols to read it
     fun viableProtocols(node: LetNode): Set<Protocol> =
-        protocolFactory.viableProtocols(node).filter {
+        node.protocol?.let { setOf(it.value) } ?: protocolFactory.viableProtocols(node).filter {
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/SelectionConstraintGenerator.kt:65]: Do we wanna check if the annotations are valid here?

Comment (rolph-recto) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/SelectionConstraintGenerator.kt:65]: I don't think so. The protocol annotation should be respected regardless of what the protocol factory says. We should treat it like programmer intent.

**Comment by rolph-recto** [comment]
> Why does this have to explicitly check if the program is fully annotated and skip selection? If it is fully annotated, we have a trivial selection which Z3 should immediately solve.

Seems wasteful. Again would be useful to skip in case we want to generate code for assignments that Z3 thinks is unsolvable.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/SelectionConstraintGenerator.kt (hunk 1 above)
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/SelectionConstraintGenerator.kt:65]: But protocol annotations should agree with what's actually implementable by the `CodeGenerators`. The lack of checks allows programmers to cause internal compiler errors and weird behavior.

**Comment by cacay** [comment]
> > Why does this have to explicitly check if the program is fully annotated and skip selection? If it is fully annotated, we have a trivial selection which Z3 should immediately solve.
> 
> Seems wasteful. Again would be useful to skip in case we want to generate code for assignments that Z3 thinks is unsolvable.

And this adds divergent paths for a very specific case where every single variable has an annotation. I think the maintenance cost of special cased code is worse than the 2ms calling Z3 adds.

**Comment by rolph-recto** [comment]
It's an `if` statement, what exactly is the maintenance cost here?

I think supporting a case where a protocol assignment is already determined and does not run through a solver at all is nice:
- It allows for generating annotated programs that you might want to run codegen on later.
- Like I mentioned above, this also allows codegen for assignments that a solver might think is unsolvable.

**Comment by cacay** [comment]
It's not just one if statement. It's an if statement and a 30 line function. Both of which aren't tested at all to top it off. Just check the coverage report. Generally, adding "divergent paths" to the control flow is bad coding practice.

**Comment by rolph-recto** [comment]
It's also bad coding practice to perform unnecessary computations that also limit functionality?

The coverage issue can be fixed. I'm just not clear why you think a fully annotated program is a very special case. I gave two reasons why skipping selection is useful. Do you just not think these are valid use cases?

**Comment by cacay** [comment]
It's worse than that. I think these two are explicitly **invalid** use cases. `ProtocolFactory` implements syntactic checks to make sure we don't assign operations to nodes that they cannot implement. As a user, if I annotate a node with an invalid protocol, I want to get an error message saying that protocol won't do. I don't wanna see an an internal compiler error. This is the same idea as how type annotations work. If the compiler sees `false : int` in a program, it doesn't go "I guess `false` is an integer now", it prints an error message.

As for selections Z3 doesn't like: if the programmer thinks a selection is valid but Z3 doesn't like it, then there is a bug in the compiler that should be fixed. Not side stepped (by annotating every variable).

**Pull request merged**